File: pack/Source.java
package common.pack;

import com.sun.xml.internal.rngom.parse.host.Base;
import common.CommonStatic;
import common.io.PackLoader;
import common.io.PackLoader.ZipDesc;
import common.io.assets.Admin.StaticPermitted;
import common.io.assets.AssetLoader;
import common.io.json.JsonClass;
import common.io.json.JsonDecoder;
import common.io.json.JsonEncoder;
import common.io.json.JsonField;
import common.pack.Context.ErrType;
import common.pack.PackData.UserPack;
import common.system.VImg;
import common.system.fake.FakeImage;
import common.system.files.FDFile;
import common.system.files.FileData;
import common.system.files.VFile;
import common.util.Data;
import common.util.anim.*;
import common.util.pack.Background;
import common.util.pack.Soul;
import common.util.stage.CastleImg;
import common.util.stage.Replay;
import common.util.stage.Stage;
import common.util.stage.StageMap;
import common.util.unit.Enemy;
import common.util.unit.Form;
import common.util.unit.Trait;
import common.util.unit.Unit;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Stream;

public abstract class Source {

	public interface AnimLoader {
		VImg getEdi();

		ImgCut getIC();

		MaAnim[] getMA();

		MaModel getMM();

		ResourceLocation getName();

		FakeImage getNum();

		int getStatus();

		VImg getUni();
	}

	public interface SourceLoader {

		FileData loadFile(BasePath base, ResourceLocation id, String str);

	}

	@JsonClass
	public static class ResourceLocation {

		public static final String LOCAL = "_local";

		@JsonField
		public String pack, id;

		@JsonField
		public BasePath base;

		@JsonClass.JCConstructor
		public ResourceLocation() {

		}

		public ResourceLocation(String pack, String id) {
			this.pack = pack;
			this.id = id;
		}

		public ResourceLocation(String pack, String id, BasePath base) {
			this.pack = pack;
			this.id = id;
			this.base = base;
		}

		public void setBase(BasePath b) {
			base = b;
		}

		@JsonClass.JCGetter
		public AnimCI getAnim() {
			if (pack.equals(LOCAL))
				return AnimCE.map().get(id);

			return UserProfile.getUserPack(pack).source.loadAnimation(id, base);
		}

		public String getPath() {
			return "./" + pack + "/" + base + "/" + id;
		}

		@JsonClass.JCGetter
		public Replay getReplay() {
			if (pack.equals(LOCAL))
				return Replay.getMap().get(id);
			Source s = UserProfile.getUserPack(pack).source;
			String path = "./" + BasePath.REPLAY + "/" + id + ".replay";
			return Data.err(() -> Replay.read(s.getFileData(path).getStream()));
		}

		@Override
		public String toString() {
			return pack + "/" + id;
		}

		@JsonDecoder.OnInjected
		public void onInjectSource() {
			Object zip = UserProfile.getStatic(UserProfile.CURRENT_PACK, () -> null);

			if (zip == null) // FIXME Check if replays of external packs aren't null
				return;

			if (this.pack.equals(LOCAL) && zip instanceof ZipSource) {
				this.pack = ((ZipSource) zip).id;
				this.id = "_mapped_" + this.id;
			}

			UserPack pack = (UserPack) UserPack.getPack(zip instanceof Workspace ? ((Workspace) zip).id : this.pack);

			if (UserProfile.isOlderPack(pack, "0.6.9.1"))
				this.base = BasePath.ANIM;
		}

	}

	@StaticPermitted
	public static class SourceAnimLoader implements Source.AnimLoader {

		public static final String IC = "imgcut.txt";
		public static final String MM = "mamodel.txt";
		public static final String[] MA_ENTITY = { "maanim_walk.txt", "maanim_idle.txt", "maanim_attack.txt", "maanim_kb.txt",
				"maanim_burrow_down.txt", "maanim_burrow_move.txt", "maanim_burrow_up.txt" };
		public static final String[] MA_SOUL = { "maanim_soul.txt" };
		public static final String SP = "sprite.png";
		public static final String EDI = "icon_display.png";
		public static final String UNI = "icon_deploy.png";

		private final ResourceLocation id;
		private final Source.SourceLoader loader;

		public SourceAnimLoader(ResourceLocation id, Source.SourceLoader loader) {
			this.id = id;
			this.loader = loader == null ? Workspace::loadAnimFile : loader;
		}

		private String[] getBaseMA() {
			if (id.base.equals(BasePath.ANIM))
				return MA_ENTITY;
			else
				return MA_SOUL;
		}

		@Override
		public VImg getEdi() {
			FileData edi = loader.loadFile(id.base, id, EDI);
			if (edi == null)
				return null;
			return new VImg(FakeImage.read(edi));
		}

		@Override
		public ImgCut getIC() {
			return ImgCut.newIns(loader.loadFile(id.base, id, IC));
		}

		@Override
		public MaAnim[] getMA() {
			MaAnim[] ans = new MaAnim[getBaseMA().length];
			for (int i = 0; i < getBaseMA().length; i++)
				ans[i] = MaAnim.newIns(loader.loadFile(id.base, id, getBaseMA()[i]));
			return ans;
		}

		@Override
		public MaModel getMM() {
			return MaModel.newIns(loader.loadFile(id.base, id, MM));
		}

		@Override
		public ResourceLocation getName() {
			return id;
		}

		@Override
		public FakeImage getNum() {
			return FakeImage.read(loader.loadFile(id.base, id, SP));
		}

		@Override
		public int getStatus() {
			return id.pack.equals("_local") ? 0 : 1;
		}

		@Override
		public VImg getUni() {
			FileData uni = loader.loadFile(id.base, id, UNI);
			if (uni == null)
				return null;
			return new VImg(FakeImage.read(uni));
		}

	}

	public static class SourceAnimSaver {

		private final ResourceLocation id;
		private final AnimCI anim;

		public SourceAnimSaver(ResourceLocation name, AnimCI animCI) {
			this.id = name;
			this.anim = animCI;
		}

		/**
		 * Delete animation
		 * @param unload If this variable is true, it means that this method is called for completely deleting process
		 */
		public void delete(boolean unload) {
			if(unload)
				anim.unload();

			CommonStatic.ctx.noticeErr(
					() -> Context.delete(CommonStatic.ctx.getWorkspaceFile("./" + id.pack + "/" + id.base + "/" + id.id)),
					ErrType.ERROR, "failed to delete animation: " + id);
		}

		public void saveAll() {
			saveData();
			saveImgs();
		}

		public void saveData() {
			try {
				write("imgcut.txt", anim.imgcut::write);
				write("mamodel.txt", anim.mamodel::write);
				if (id.base.equals(BasePath.ANIM)) {
					for (int i = 0; i < SourceAnimLoader.MA_ENTITY.length; i++) {
						write(SourceAnimLoader.MA_ENTITY[i], anim.anims[i]::write);
					}
				} else {
					write(SourceAnimLoader.MA_SOUL[0], anim.anims[0]::write);
				}
			} catch (IOException e) {
				CommonStatic.ctx.noticeErr(e, ErrType.ERROR, "Error during saving animation data: " + anim);
			}
		}

		public void saveIconDeploy() {
			if (anim.getUni() != null && !id.base.equals(BasePath.SOUL))
				CommonStatic.ctx.noticeErr(() -> write("icon_deploy.png", anim.getUni().getImg()), ErrType.ERROR,
						"Error during saving deploy icon: " + id);
		}

		public void saveIconDisplay() {
			if (anim.getEdi() != null)
				CommonStatic.ctx.noticeErr(() -> write("icon_display.png", anim.getEdi().getImg()), ErrType.ERROR,
						"Error during saving display icon: " + id);
		}

		public void saveImgs() {
			saveSprite();
			saveIconDisplay();
			saveIconDeploy();
		}

		public void saveSprite() {
			CommonStatic.ctx.noticeErr(() -> write("sprite.png", anim.getNum()), ErrType.ERROR,
					"Error during saving sprite sheet: " + id);
		}

		private void write(String type, Consumer<PrintStream> con) throws IOException {
			File f = CommonStatic.ctx.getWorkspaceFile(id.getPath() + "/" + type);
			Context.check(f);
			PrintStream ps = new PrintStream(f, StandardCharsets.UTF_8.toString());
			con.accept(ps);
			ps.close();
		}

		private void write(String type, FakeImage img) throws IOException {
			File f = CommonStatic.ctx.getWorkspaceFile(id.getPath() + "/" + type);
			Context.check(f);
			Context.check(FakeImage.write(img, "PNG", f), "save", f);
		}

	}

	public static class Workspace extends Source {

		public static void loadAnimations(String id) {
			if (id == null)
				id = ResourceLocation.LOCAL;
			File animFolder = CommonStatic.ctx.getWorkspaceFile("./" + id + "/" + BasePath.ANIM + "/");
			File soulFolder = CommonStatic.ctx.getWorkspaceFile("./" + id + "/" + BasePath.SOUL + "/");
			if (animFolder.exists() && animFolder.isDirectory()) {
				File[] animFiles = animFolder.listFiles();
				Arrays.sort(animFiles);
				for (File f : animFiles) {
					String path = "./" + id + "/" + BasePath.ANIM + "/" + f.getName() + "/sprite.png";

					if (f.isDirectory() && CommonStatic.ctx.getWorkspaceFile(path).exists()) {
						ResourceLocation rl = new ResourceLocation(id, f.getName(), Source.BasePath.ANIM);
						AnimCE anim = new AnimCE(rl);

						AnimCE.map().put(f.getName(), anim);
					}
				}
			}
			if (soulFolder.exists() && soulFolder.isDirectory()) {
				File[] soulFiles = soulFolder.listFiles();
				Arrays.sort(soulFiles);
				for (File f : soulFiles) {
					String path = "./" + id + "/" + BasePath.SOUL + "/" + f.getName() + "/sprite.png";

					if (f.isDirectory() && CommonStatic.ctx.getWorkspaceFile(path).exists()) {
						ResourceLocation rl = new ResourceLocation(id, f.getName(), BasePath.SOUL);
						AnimCE anim = new AnimCE(rl);

						AnimCE.map().put(f.getName(), anim);
					}
				}
			}
		}

		public static void autoSave() {
			AnimCE.map().values().forEach(AnimCE::autosave);
			for (UserPack up : UserProfile.getUserPacks())
				if (up.source instanceof Workspace)
					CommonStatic.ctx.noticeErr(() -> ((Workspace) up.source).save(up, true), ErrType.WARN,
							"failed to save pack " + up.desc.names.toString());
		}

		public static void saveLocalAnimations() {
			AnimCE.map().values().forEach(AnimCE::save);
		}

		public static void saveWorkspace() {
			for (UserPack up : UserProfile.getUserPacks())
				if (up.source instanceof Workspace)
					CommonStatic.ctx.noticeErr(() -> ((Workspace) up.source).save(up, false), ErrType.WARN,
							"failed to save pack " + up.desc.names.toString());
		}

		public static void validate(ResourceLocation rl) {
			String id = rl.id;
			int num = 0;
			while (CommonStatic.ctx.getWorkspaceFile(rl.getPath()).exists())
				rl.id = id + "_" + (num++);
		}

		public static String validateString(String str) {
			if (str == null || str.length() == 0)
				str = "no_name";
			str = str.replaceAll("[^0-9a-z_]", "_");
			if (str.charAt(0) < '0')
				str = "a_" + str;
			return str;
		}

		public static String validateWorkspace(String str) {
			String id = validateString(str);
			int num = 0;
			while (CommonStatic.ctx.getWorkspaceFile("./" + str).exists())
				str = id + "_" + (num++);
			return id;
		}

		public static String generatePackID() {
			String format = "abcdefghijklmnopqrstuvwxyz0123456789";
			Random random = new Random();

			StringBuilder result = new StringBuilder();

			while (result.length() < 8) {
				char ch = format.charAt((int) (random.nextFloat() * format.length()));

				result.append(ch);
			}

			return result.toString();
		}

		private static FileData loadAnimFile(BasePath base, ResourceLocation id, String str) {
			String path = "./" + id.pack + "/" + base.toString() + "/" + id.id + "/" + str;
			File f = CommonStatic.ctx.getWorkspaceFile(path);
			if (!f.exists())
				return null;
			return new FDFile(f);
		}

		public Workspace(String id) {
			super(id);
		}

		@Override
		public void delete() {
			try {
				Context.delete(getFile(""));
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

		public void export(UserPack pack, String password, String parentPassword, Consumer<Double> prog) throws Exception {
			ArrayList<AnimCE> anims = new ArrayList<>();

			for (Enemy e : pack.enemies) {
				AnimCE anim = (AnimCE) e.anim;
				if (anim.id.pack.equals(ResourceLocation.LOCAL)) {
					if(!anims.contains(anim)) {
						anims.add(anim);
					} else {
						anim.id.pack = ResourceLocation.LOCAL;
						anim.id.id = anim.id.id.replaceAll("^_mapped_", "");
					}

					new SourceAnimSaver(new ResourceLocation(pack.getSID(), "_mapped_"+anim.id.id, anim.id.base), anim).saveAll();

					anim.id.pack = pack.getSID();
					anim.id.id = "_mapped_"+anim.id.id;
				}
				if (anim.id.pack.startsWith(".temp_"))
					anim.id.pack = anim.id.pack.substring(6);
			}
			for (Unit u : pack.units)
				for (Form f : u.forms) {
					AnimCE anim = (AnimCE) f.anim;
					if (anim.id.pack.equals(ResourceLocation.LOCAL)) {
						if(!anims.contains(anim)) {
							anims.add(anim);
						} else {
							anim.id.pack = ResourceLocation.LOCAL;
							anim.id.id = anim.id.id.replaceAll("^_mapped_", "");
						}

						new SourceAnimSaver(new ResourceLocation(pack.getSID(), "_mapped_"+anim.id.id, anim.id.base), anim).saveAll();

						anim.id.pack = pack.getSID();
						anim.id.id = "_mapped_"+anim.id.id;
					}
					if (anim.id.pack.startsWith(".temp_"))
						anim.id.pack = anim.id.pack.substring(6);
				}
			for (Soul s : pack.souls) {
				AnimCE anim = (AnimCE) s.anim;
				if (anim.id.pack.equals(ResourceLocation.LOCAL)) {
					if(!anims.contains(anim)) {
						anims.add(anim);
					} else {
						anim.id.pack = ResourceLocation.LOCAL;
						anim.id.id = anim.id.id.replaceAll("^_mapped_", "");
					}

					new SourceAnimSaver(new ResourceLocation(pack.getSID(), "_mapped_"+anim.id.id, anim.id.base), anim).saveAll();

					anim.id.pack = pack.getSID();
					anim.id.id = "_mapped_"+anim.id.id;
				}
				if (anim.id.pack.startsWith(".temp_"))
					anim.id.pack = anim.id.pack.substring(6);
			}
			for (StageMap sm : pack.mc.maps)
				for (Stage st : sm.list)
					for (Replay rep : st.recd)
						if (rep != null && rep.rl.pack.startsWith(".temp_")) {
							rep.rl.pack = rep.rl.pack.substring(6);
						}
			save(pack, false);
			String star = id.startsWith(".temp_") ? "./packs/" : "./exports/";
			File tar = CommonStatic.ctx.getAuxFile(star + pack.getSID() + ".pack.bcuzip");
			File dst = CommonStatic.ctx.getAuxFile(star + ".pack.bcuzip.temp");
			File src = CommonStatic.ctx.getWorkspaceFile("./" + id);
			if (tar.exists())
				Context.delete(tar);
			Context.check(dst);

			PackData.PackDesc desc = pack.desc.clone();

			if (parentPassword != null) {
				desc.parentPassword = PackLoader.getMD5(parentPassword.getBytes(StandardCharsets.UTF_8), 16);
			} else {
				desc.parentPassword = null;
			}

			PackLoader.writePack(dst, src, desc, password, prog);
			Context.renameTo(dst, tar);

			for(AnimCE anim : anims) {
				anim.id.pack = ResourceLocation.LOCAL;
				anim.id.id = anim.id.id.replaceAll("^_mapped_", "");
			}
		}

		public File getBGFile(Identifier<Background> id) {
			return getFile("./" + BasePath.BG.toString() + "/" + Data.trio(id.id) + ".png");
		}

		public File getCasFile(Identifier<CastleImg> id) {
			return getFile("./" + BasePath.CASTLE + "/" + Data.trio(id.id) + ".png");
		}

		public File getTraitIconFile(Identifier<Trait> id) {
			return getFile("./" + BasePath.TRAIT + "/" + Data.trio(id.id) + ".png");
		}

		@Override
		public FileData getFileData(String string) {
			return new FDFile(getFile(string));
		}

		@Override
		public String[] listFile(String path) {
			return getFile(path).list();
		}

		@Override
		public AnimCE loadAnimation(String name, BasePath base) {
			return new AnimCE(new ResourceLocation(id, name, base));
		}

		@Override
		public VImg readImage(String path, int ind) {
			VFile vf = VFile.getFile(getFile(path + "/" + Data.trio(ind) + ".png"));

			if(vf == null) {
				return null;
			} else {
				return new VImg(vf);
			}
		}

		@Override
		public InputStream streamFile(String path) throws IOException {
			return new FileInputStream(getFile(path));
		}

		public OutputStream writeFile(String path) throws IOException {
			File f = getFile(path);
			Context.check(f);
			return new FileOutputStream(f);
		}

		protected void save(UserPack up, boolean auto) throws IOException {
			File f = auto ? CommonStatic.ctx.getWorkspaceFile("./_autosave/pack_" + id + ".json") : getFile("pack.json");
			Context.check(f);
			OutputStreamWriter fw = new OutputStreamWriter(new FileOutputStream(f), StandardCharsets.UTF_8);
			fw.write(JsonEncoder.encode(up).toString());
			fw.close();
		}

		private File getFile(String path) {
			return CommonStatic.ctx.getWorkspaceFile("./" + id + "/" + path);
		}

	}

	public static class ZipSource extends Source {

		public final ZipDesc zip;

		public ZipSource(ZipDesc desc) {
			super(desc.desc.id);
			zip = desc;
		}

		@Override
		public void delete() {
			zip.delete();
		}

		@Override
		public FileData getFileData(String string) {
			return zip.tree.find(string).getData();
		}

		public File getPackFile() {
			return zip.getZipFile();
		}

		@Override
		public String[] listFile(String path) {
			VFile dir = zip.tree.find(path);
			if (dir == null)
				return null;
			Collection<VFile> col = dir.list();
			if (col == null)
				return null;
			String[] ans = new String[col.size()];
			int i = 0;
			for (VFile vf : col) {
				ans[i++] = vf.name;
			}
			return ans;
		}

		@Override
		public AnimCI loadAnimation(String name, BasePath base) {
			return new AnimCI(new SourceAnimLoader(new ResourceLocation(id, name, base), this::loadAnimationFile));
		}

		@Override
		public VImg readImage(String path, int ind) {
			String fullPath = path.startsWith("./") ? path + "/" + Data.trio(ind) + ".png" : "./" + path + "/" + Data.trio(ind) + ".png";
			VFile vf = zip.tree.find(fullPath);

			if (vf == null)
				return null;
			return new VImg(vf);
		}

		@Override
		public InputStream streamFile(String path) throws Exception {
			return zip.readFile(path);
		}

		public Workspace unzip(String password, Consumer<Double> prog) throws Exception {
			if (!zip.match(PackLoader.getMD5(password.getBytes(StandardCharsets.UTF_8), 16)))
				return null;
			File f = CommonStatic.ctx.getWorkspaceFile("./" + id + "/pack.json");
			File folder = f.getParentFile();
			if (folder.exists()) {
				if (!CommonStatic.ctx.confirmDelete(f))
					return null;
				Context.delete(f);
			}
			if (!folder.exists())
				Context.check(folder.mkdirs(), "create", folder);
			if (!f.exists())
				Context.check(f.createNewFile(), "create", f);
			Workspace ans = new Workspace(id);
			zip.unzip(id -> {
				File file = ans.getFile(id);
				Context.check(file);
				return file;
			}, prog);
			return ans;
		}

		private FileData loadAnimationFile(BasePath base, ResourceLocation id, String path) {
			VFile vf = zip.tree.find("./" + base.toString() + "/" + id.id + "/" + path);
			return vf == null ? null : vf.getData();
		}

	}

	public enum BasePath {
		ANIM("animations"),
		BG("backgrounds"),
		CASTLE("castles"),
		MUSIC("musics"),
		REPLAY("replays"),
		SOUL("souls"),
		TRAIT("traitIcons");

		private final String path;

		BasePath(String str) {
			path = str;
		}

		public String toString() {
			return path;
		}

	}

	public final String id;

	public Source(String id) {
		this.id = id;
	}

	public abstract void delete();

	public abstract FileData getFileData(String string);

	public abstract String[] listFile(String path);

	public abstract AnimCI loadAnimation(String name, BasePath base);

	/**
	 * read images from file. Use it
	 */
	//TODO: might be able to use BasePath for path
	public abstract VImg readImage(String path, int ind);

	/**
	 * used for streaming music. Do not use it for images and small text files
	 */
	public abstract InputStream streamFile(String path) throws Exception;

}

File: pack/Source.kt
package common.pack

import common.CommonStatic
import common.io.PackLoader
import common.io.PackLoader.PatchFile
import common.io.PackLoader.ZipDesc
import common.io.PackLoader.ZipDesc.FileDesc
import common.io.assets.Admin.StaticPermitted
import common.pack.Context.ErrType
import common.pack.Context.SupExc
import common.pack.Sourceimport
import common.system.VImg
import common.system.fake.FakeImage
import common.system.fake.ImageBuilder
import common.system.files.FDFile
import common.system.files.FileData
import common.util.Data
import common.util.anim.AnimCE
import common.util.anim.AnimCI
import common.util.anim.ImgCut
import common.util.anim.MaAnim
import common.util.anim.MaModel
import common.util.pack.Background
import common.util.stage.CastleImg
import java.io.*
import java.util.*
import java.util.function.Consumer

com.google.api.client.json.jackson2.JacksonFactory
import kotlin.Throws
import java.io.IOException
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.PrintStream

abstract class Source(val id: String) {
    interface AnimLoader {
        fun getEdi(): VImg?
        fun getIC(): ImgCut
        fun getMA(): Array<MaAnim?>
        fun getMM(): MaModel
        fun getName(): ResourceLocation
        fun getNum(): FakeImage?
        fun getStatus(): Int
        fun getUni(): VImg?
    }

    class ResourceLocation(var pack: String, var id: String) {
        companion object {
            const val LOCAL = "_local"
        }
    }

    @StaticPermitted
    class SourceAnimLoader(private val id: ResourceLocation, loader: Source.SourceAnimLoader.SourceLoader?) : AnimLoader {
        interface SourceLoader {
            fun loadFile(id: ResourceLocation?, str: String?): FileData?
        }

        private val loader: Source.SourceAnimLoader.SourceLoader
        override fun getEdi(): VImg? {
            val edi = loader.loadFile(id, EDI) ?: return null
            return CommonStatic.ctx.noticeErr<VImg>(SupExc<VImg> { VImg(FakeImage.Companion.read(edi)) }, ErrType.ERROR,
                    "failed to read Display Icon$id")
        }

        override fun getIC(): ImgCut {
            return ImgCut.Companion.newIns(loader.loadFile(id, IC))
        }

        override fun getMA(): Array<MaAnim?> {
            val ans: Array<MaAnim?> = arrayOfNulls<MaAnim>(MA.size)
            for (i in MA.indices) ans[i] = MaAnim.Companion.newIns(loader.loadFile(id, MA[i]))
            return ans
        }

        override fun getMM(): MaModel {
            return MaModel.Companion.newIns(loader.loadFile(id, MM))
        }

        override fun getName(): ResourceLocation {
            return id
        }

        override fun getNum(): FakeImage? {
            return CommonStatic.ctx.noticeErr<FakeImage>(SupExc<FakeImage> { FakeImage.Companion.read(loader.loadFile(id, SP)) }, ErrType.ERROR,
                    "failed to read sprite sheet $id")
        }

        override fun getStatus(): Int {
            return if (id.pack == "_local") 0 else 1
        }

        override fun getUni(): VImg? {
            val uni = loader.loadFile(id, UNI) ?: return null
            return CommonStatic.ctx.noticeErr<VImg>(SupExc<VImg> { VImg(FakeImage.Companion.read(uni)) }, ErrType.ERROR,
                    "failed to read deploy icon $id")
        }

        companion object {
            const val IC = "imgcut.txt"
            const val MM = "mamodel.txt"
            val MA = arrayOf("maanim_walk.txt", "maanim_idle.txt", "maanim_attack.txt", "maanim_kb.txt",
                    "maanim_burrow_down.txt", "maanim_burrow_move.txt", "maanim_burrow_up.txt")
            const val SP = "sprite.png"
            const val EDI = "icon_display.png"
            const val UNI = "icon_deploy.png"
        }

        init {
            this.loader = loader
                    ?: Source.SourceAnimLoader.SourceLoader { id: ResourceLocation, str: String -> Workspace.loadAnimFile(id, str) }
        }
    }

    class SourceAnimSaver(private val id: ResourceLocation, animCI: AnimCI) {
        private val anim: AnimCI
        fun delete() {
            CommonStatic.ctx.noticeErr(
                    Context.RunExc { Context.Companion.delete(CommonStatic.ctx.getWorkspaceFile("./" + id.pack + "/animations/" + id.id)) },
                    ErrType.ERROR, "failed to delete animation: $id")
        }

        fun saveAll() {
            saveData()
            saveImgs()
        }

        fun saveData() {
            try {
                write("imgcut.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.imgcut.write(ps) })
                write("mamodel.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.mamodel.write(ps) })
                write("maanim_walk.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.anims.get(0).write(ps) })
                write("maanim_idle.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.anims.get(1).write(ps) })
                write("maanim_attack.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.anims.get(2).write(ps) })
                write("maanim_kb.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.anims.get(3).write(ps) })
                write("maanim_burrow_down.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.anims.get(4).write(ps) })
                write("maanim_burrow_move.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.anims.get(5).write(ps) })
                write("maanim_burrow_up.txt", Consumer<PrintStream> { ps: PrintStream? -> anim.anims.get(6).write(ps) })
            } catch (e: IOException) {
                CommonStatic.ctx.noticeErr(e, ErrType.ERROR, "Error during saving animation data: $anim")
            }
        }

        fun saveIconDeploy() {
            if (anim.getUni() != null) CommonStatic.ctx.noticeErr(Context.RunExc { write("icon_deploy.png", anim.getUni().getImg()) }, ErrType.ERROR,
                    "Error during saving deploy icon: $id")
        }

        fun saveIconDisplay() {
            if (anim.getEdi() != null) CommonStatic.ctx.noticeErr(Context.RunExc { write("icon_display.png", anim.getEdi().getImg()) }, ErrType.ERROR,
                    "Error during saving display icon: $id")
        }

        fun saveImgs() {
            saveSprite()
            saveIconDisplay()
            saveIconDeploy()
        }

        fun saveSprite() {
            CommonStatic.ctx.noticeErr(Context.RunExc { write("sprite.png", anim.getNum()) }, ErrType.ERROR,
                    "Error during saving sprite sheet: $id")
        }

        @Throws(IOException::class)
        private fun write(type: String, con: Consumer<PrintStream>) {
            val f: File = CommonStatic.ctx.getWorkspaceFile("./" + id.pack + "/animations/" + id.id + "/" + type)
            Context.Companion.check(f)
            val ps = PrintStream(f)
            con.accept(ps)
            ps.close()
        }

        @Throws(IOException::class)
        private fun write(type: String, img: FakeImage) {
            val f: File = CommonStatic.ctx.getWorkspaceFile("./" + id.pack + "/animations/" + id.id + "/" + type)
            Context.Companion.check(f)
            Context.Companion.check(FakeImage.Companion.write(img, "PNG", f), "save", f)
        }

        init {
            anim = animCI
        }
    }

    class Workspace(id: String?) : Source(id) {
        fun getBGFile(id: PackData.Identifier<Background?>): File {
            return getFile("./backgrounds/" + Data.Companion.trio(id.id) + ".png")
        }

        fun getCasFile(id: PackData.Identifier<CastleImg?>): File {
            return getFile("./castles/" + Data.Companion.trio(id.id) + ".png")
        }

        override fun loadAnimation(name: String?): AnimCI? {
            return AnimCI(SourceAnimLoader(ResourceLocation(id, name), null))
        }

        @Throws(IOException::class)
        override fun readImage(path: String): FakeImage? {
            return ImageBuilder.Companion.builder.build(getFile(path))
        }

        @Throws(IOException::class)
        override fun streamFile(path: String): InputStream? {
            return FileInputStream(getFile(path))
        }

        @Throws(IOException::class)
        fun writeFile(path: String): OutputStream {
            val f = getFile(path)
            Context.Companion.check(f)
            return FileOutputStream(f)
        }

        private fun getFile(path: String): File {
            return CommonStatic.ctx.getWorkspaceFile("./$id/$path")
        }

        companion object {
            fun loadAnimations(id: String?): List<AnimCE> {
                var id = id
                if (id == null) id = ResourceLocation.Companion.LOCAL
                val folder: File = CommonStatic.ctx.getWorkspaceFile("./$id/animations/")
                val list: MutableList<AnimCE> = ArrayList<AnimCE>()
                if (!folder.exists() || !folder.isDirectory) return list
                for (f in folder.listFiles()) {
                    val path = "./" + id + "/animations/" + f.name + "/sprite.png"
                    if (f.isDirectory && CommonStatic.ctx.getWorkspaceFile(path).exists()) list.add(AnimCE(ResourceLocation(id, f.name)))
                }
                return list
            }

            fun validate(rl: ResourceLocation): ResourceLocation {
                // FIXME find valid path
                return rl
            }

            fun loadAnimFile(id: ResourceLocation, str: String): FileData {
                return FDFile(CommonStatic.ctx.getWorkspaceFile("./" + id.pack + "/animations/" + id.id + "/" + str))
            }
        }
    }

    class ZipSource(desc: ZipDesc) : Source(desc.desc.id) {
        private val zip: ZipDesc
        override fun loadAnimation(name: String?): AnimCI? {
            return AnimCI(SourceAnimLoader(ResourceLocation(id, name)) { id: ResourceLocation, path: String -> loadAnimationFile(id, path) })
        }

        @Throws(Exception::class)
        override fun readImage(path: String): FakeImage? {
            return zip.tree.find(path).getData().getImg()
        }

        @Throws(Exception::class)
        override fun streamFile(path: String): InputStream? {
            return zip.readFile(path)
        }

        @Throws(Exception::class)
        fun unzip(password: String): Workspace? {
            if (!zip.match(PackLoader.getMD5(password.toByteArray(), 16))) return null
            val f: File = CommonStatic.ctx.getWorkspaceFile("./$id/main.pack.json")
            val folder = f.parentFile
            if (folder.exists()) {
                if (!CommonStatic.ctx.confirmDelete()) return null
                Context.Companion.delete(f)
            }
            Context.Companion.check(folder.mkdirs(), "create", folder)
            Context.Companion.check(f.createNewFile(), "create", f)
            val ans = Workspace(id)
            zip.unzip(PatchFile { id: String? ->
                val file = ans.getFile(id)
                Context.Companion.check(file.createNewFile(), "create", file)
                file
            })
            return ans
        }

        private fun loadAnimationFile(id: ResourceLocation, path: String): FileDesc {
            return zip.tree.find("./animations/" + id.id + "/" + path).getData()
        }

        init {
            zip = desc
        }
    }

    abstract fun loadAnimation(name: String?): AnimCI?

    /** read images from file. Use it  */
    @Throws(Exception::class)
    abstract fun readImage(path: String): FakeImage?

    /** used for streaming music. Do not use it for images and small text files  */
    @Throws(Exception::class)
    abstract fun streamFile(path: String): InputStream?
}
